src/file.py

```diff
@@ -0,0 +1,406 @@
+            )
+            raise AuthenticationException(common.ACCESS_ERROR)
+        if res.status.code != cs3code.CODE_OK:
+            self.log.error(
+                'msg="Failed to initiateFileUpload on touch" filepath="%s" trace="%s" code="%s" reason="%s"'
+                % (resource.file_ref, res.status.trace, res.status.code, res.status.message.replace('"', "'"))
+            )
+            raise IOError(res.status.message)
+        self.log.debug(f'msg="Invoked touchfile" result="{res}"')
+
+    def write_file(self, resource: Resource, content: Union[str, bytes], size: int, islock=False) -> None:
+        """Write a file using the given userid as access token. The entire content is written
+        and any pre-existing file is deleted (or moved to the previous version if supported).
+        The islock flag is currently not supported. The backend should at least support
+        writing the file with O_CREAT|O_EXCL flags to prevent races."""
+        tstart = time.time()
+        # prepare endpoint
+        if size == -1:
+            if isinstance(content, str):
+                content = bytes(content, "UTF-8")
+            size = len(content)
+        req = cs3sp.InitiateFileUploadRequest(
+            ref=resource.ref,
+            opaque=types.Opaque(map={"Upload-Length": types.OpaqueEntry(decoder="plain", value=str.encode(str(size)))}),
+        )
+        res = self.gateway.InitiateFileUpload(request=req, metadata=[("x-access-token", self.auth.token)])
+        if res.status.code == cs3code.CODE_FAILED_PRECONDITION:
+            self.log.info(
+                'msg="Lock mismatch uploading file" filepath="%s" reason="%s"'
+                % (self.auth.token, res.status.message.replace('"', "'"))
+            )
+            raise IOError(common.EXCL_ERROR)
+        if res.status.code == cs3code.CODE_UNAUTHENTICATED:
+            self.log.error(
+                'msg="Authentication failed on write" filepath="%s" userid="%s" trace="%s" reason="%s"'
+                % (resource.file_ref, self.auth.token[-20:], res.status.trace, res.status.message.replace('"', "'"))
+            )
+            raise AuthenticationException(common.ACCESS_ERROR)
+        if res.status.code != cs3code.CODE_OK:
+            self.log.error(
+                'msg="Failed to initiateFileUpload on write" filepath="%s" trace="%s" code="%s" reason="%s"'
+                % (resource.file_ref, res.status.trace, res.status.code, res.status.message.replace('"', "'"))
+            )
+            raise IOError(res.status.message)
+        tend = time.time()
+        self.log.debug(
+            'msg="writefile: InitiateFileUploadRes returned" trace="%s" protocols="%s"'
+            % (res.status.trace, res.protocols)
+        )
+
+        # Upload
+        try:
+            protocol = [p for p in res.protocols if p.protocol in ["simple", "spaces"]][0]
```

Comment:
I'll look into this

Reply:
That is correct, will add this

Reply:
cc @glpatcern

Reply:
I wonder what happens if server offers `tus` and client does not implement it - the current implementation in wopiserver does not implement TUS. This will have to be tested.

Reply:
I have made it now so that tus_enabled can be set in the config